File: R/Consumer.R
#' Consumer reference class
#'
#' @export
Consumer <- R6Class("Consumer",
    public = list(
        #' @description
        #' Create a new consumer instance
        #'
        #' @param config (`list()`)\cr
        #'   list of config values. These are passed without modification to
        #'   librdkafka
        initialize = function(config) {
            consumer_rcpp_module <- Rcpp::Module("consumer_module")
            private$consumer <- new(consumer_rcpp_module$Consumer, config)
        },
        #' @description
        #' Subscribe to topic
        #'
        #' @param topic (`character`)\cr
        #'   kafka topic
        subscribe = function(topic) {
            private$consumer$subscribe(topic)
        },
        #' @description
        #' Consume single message
        #'
        #' @param timeout (`integer`)\cr
        #'   timeout in ms.
        consume = function(timeout = 1000) {
            result <- private$consumer$consume(timeout)

            message <- new_kafka_message(
                value = result$payload_data,
                key = result$key
            )

            new_kafka_result(
                message,
                result$error_code,
                result$error_message
            )
        },
        #' @description
        #' Commit current offset
        #'
        #' @param async (`logical`)\cr
        #'   commit async
        commit = function(async = FALSE) {
            private$consumer$commit(async)
        },
        #' @description
        #' Unsubscribe from current topic
        unsubscribe = function() {
            private$consumer$unsubscribe()
        },
        #' @description
        #' Close consumer
        close = function() {
            private$consumer$close()
        }
    ),
    private = list(
        consumer = NULL
    )
)

File: src/consumer.cpp
#include <librdkafka/rdkafkacpp.h>
#include <Rcpp.h>
#include "utils.h"
#include <iostream>

using namespace Rcpp;

class Consumer
{
public:
    Consumer(Rcpp::List conf_)
    {
        std::string errstr;

        conf = generate_kafka_config(conf_);

        consumer = RdKafka::KafkaConsumer::create(conf, errstr);

        if (!consumer)
        {
            Rcpp::stop("Failed to create Kafka Consumer: " + errstr);
        }
    }

    void subscribe(std::vector<std::string> topics)
    {
        RdKafka::ErrorCode err = consumer->subscribe(topics);

        if (err != RdKafka::ERR_NO_ERROR)
        {
            Rcpp::stop("Failed to subscribe to topics: " + RdKafka::err2str(err));
        }
    }

    void unsubscribe()
    {
        std::vector<std::string> topics;
        consumer->subscription(topics);

        for (const auto &topic : topics)
        {
            Rcpp::Rcout << "Unsubscribing from " << topic << std::endl;
        }
        RdKafka::ErrorCode err = consumer->unsubscribe();

        if (err != RdKafka::ERR_NO_ERROR)
        {
            Rcpp::stop("Failed to unsubscribe: " + RdKafka::err2str(err));
        }
    }

    void commit(bool async = false)
    {
      RdKafka::ErrorCode err;
      Rcpp::Rcout << "committing message" << std::endl;
      if (async) {
         err = consumer->commitAsync();
      }

      else {
         err = consumer->commitSync();
      }

        if (err != RdKafka::ERR_NO_ERROR)
        {
            Rcpp::stop("Failed commit: " + RdKafka::err2str(err));
        }
    }

    void close()
    {
        Rcpp::Rcout << "Closing consumer" << std::endl;
        RdKafka::ErrorCode err = consumer->close();
        if (err != RdKafka::ERR_NO_ERROR)
        {
            Rcpp::stop("Failed to close subscriber: " + RdKafka::err2str(err));
        }
    }

    List consume(const int timeout_ms)
    {
        List result;

        const void *payload;
        size_t payload_size;

        RdKafka::Message *msg = consumer->consume(timeout_ms);
        std::cout << msg << std::endl;

        switch (msg->err())
        {
        case RdKafka::ERR_NO_ERROR:
            payload = msg->payload();
            payload_size = msg->len();

            result["payload_data"] = std::string(static_cast<const char *>(payload), payload_size);

            if (msg->key()) {
                result["key"] = *msg->key();
            }
            
            break;
        case RdKafka::ERR__TIMED_OUT:
            break;
        default:
            result["error_message"] = msg->errstr();
            result["error_code"] = RdKafka::err2str(msg->err());
            Rcpp::Rcout << "Error: " << msg->errstr() << std::endl;
        }

        return result;
    }

private:
    RdKafka::Conf *conf;
    RdKafka::KafkaConsumer *consumer;
};

RCPP_MODULE(consumer_module)
{
    class_<Consumer>("Consumer")
        .constructor<Rcpp::List>()
        .method("subscribe", &Consumer::subscribe)
        .method("unsubscribe", &Consumer::unsubscribe)
        .method("commit", &Consumer::commit)
        .method("consume", &Consumer::consume)
        .method("close", &Consumer::close);
}
